### PR TITLE
[Learn] Automation API Pathway

### PR DESCRIPTION
**Pull request by nimbinatus**
Adding the Automation API Pathway.

closes pulumi/devrel-team#89

### CODE REVIEW

**Comment by nimbinatus** [comment]
Note: My Falcon API went a bit sideways for the last module. That will go up later tonight.

**Comment by cnunciato** [comment]
I'm still reviewing this in detail, but in the meantime, a few things to think about:

* The intro could probably do more to explain the problem to be solved, so the overall shape of the solution is clear and well motivated (i.e., "why Automation API?") Without this, I had trouble picturing what we'd be doing and why. A concrete scenario and a diagram could go a long way here, I think -- just a little something more to give me a stronger mental model of what the situation is, what the parts are, what each one's for, and how each step brings me closer to the ultimate goal.

* The directory structure and filenames should probably be spelled out more clearly. I'm definitely not strong with Python (working on it!), but I did get confused as to which files were supposed to go where, what they were supposed to be called, etc. (The first thing I did was run `pulumi new` in `learn-auto-api`, but at some point I got the sense that maybe wasn't what you'd intended, so I had to go back and do a bunch of cleanup up, renaming, debugging, etc., to get back on the right track, and I never quite made it.)

* I get the choice of the Random provider (keeps it simple), but for me at least, it was a bit of a mental shift to think about "random string" as a placeholder for "some theoretical cloud infra", especially since was already making another mental shift to think about driving Pulumi programmatically. I'm not sure what the best alternative is, though -- maybe Docker? Maybe it's okay -- it certainly _works_ with the Random provider -- I just kept having to remind myself that that code wasn't there just to generate a random string, it was there as a stand-in for something else.

Hope this helps as an initial pass -- lemme know what you think.

**Comment by nimbinatus** [comment]
> I'm still reviewing this in detail, but in the meantime, a few things to think about:
> 
> * The intro could probably do more to explain the problem to be solved, so the overall shape of the solution is clear and well motivated (i.e., "why Automation API?") Without this, I had trouble picturing what we'd be doing and why. A concrete scenario and a diagram could go a long way here, I think -- just a little something more to give me a stronger mental model of what the situation is, what the parts are, what each one's for, and how each step brings me closer to the ultimate goal.

Fair point. I was kinda taking it people are coming to this doc already knowing something about it, but I don't have the insight yet into who is coming from where.

> * The directory structure and filenames should probably be spelled out more clearly. I'm definitely not strong with Python (working on it!), but I did get confused as to which files were supposed to go where, what they were supposed to be called, etc. (The first thing I did was run `pulumi new` in `learn-auto-api`, but at some point I got the sense that maybe wasn't what you'd intended, so I had to go back and do a bunch of cleanup, renaming, debugging, etc., to get back on track, and I never quite made it.)

Ah, good callout. I'm actually debating whether a running file tree might be a good addition to all of these pages, to be honest. It's really hard not to have a gigantic block of code every few lines and still keep everyone on board.

> * I get the choice of the Random provider (keeps it simple), but for me at least, it was a bit of a mental shift to think about "random string" as a placeholder for "some theoretical cloud infra", especially since was already making another mental shift to think about driving Pulumi programmatically. I'm not sure what the best alternative is, though -- maybe Docker? Maybe it's okay -- it certainly _works_ with the Random provider -- I just kept having to remind myself that that code wasn't there just to generate a random string, it was there as a stand-in for something else.

Yeah, I was trying to avoid the chaos we had with the Docker provider in Fundamentals with M1 and other oddities, but I agree it might be way too theoretical. I'll have to try to make something work, maybe as a follow-up :/

> Hope this helps as an initial pass! Lemme know what you think.

It does! Thank you!

**Comment by cnunciato** [comment]
> I'm actually debating whether a running file tree might be a good addition to all of these pages, to be honest.

Glad you mentioned this! Definitely something to keep talking about as we put these pages to actual use. We have a ton of flexibility, here, so if we want to do more, we totally can.

File: themes/default/content/learn/embedding-pulumi/building-api/index.md
Comment (katcosgrove) [inline themes/default/content/learn/embedding-pulumi/building-api/index.md:30]:
```suggestion
Let’s start out with scaffolding the code by importing the right libraries and
```

File: themes/default/content/learn/embedding-pulumi/building-api/index.md
Comment (katcosgrove) [inline themes/default/content/learn/embedding-pulumi/building-api/index.md:32]:
```suggestion
refresh, destroy, and update) to code with the automation API library:
```

File: themes/default/content/learn/embedding-pulumi/building-api/index.md
Comment (katcosgrove) [inline themes/default/content/learn/embedding-pulumi/building-api/index.md:86]:
```suggestion
with the Automation API: a procedure for creating, updating, or deleting a
```

File: themes/default/content/learn/embedding-pulumi/considerations/index.md
Comment (katcosgrove) [inline themes/default/content/learn/embedding-pulumi/considerations/index.md:37]:
```suggestion
`pulumi.<log-level>`. If you’re needing audit logging (tracking which system
```

File: themes/default/content/learn/embedding-pulumi/using-the-api/index.md
Comment (katcosgrove) [inline themes/default/content/learn/embedding-pulumi/using-the-api/index.md:215]: Should "curling" be changed to "CURLing"? Probably a stylistic choice, doesn't really matter

File: themes/default/content/learn/embedding-pulumi/using-the-api/index.md
Comment (katcosgrove) [inline themes/default/content/learn/embedding-pulumi/using-the-api/index.md:287]: Missing backticks to fence codeblocks here and just below

**Review by katcosgrove — CHANGES_REQUESTED** [review]
Just a few grammatical issues!

**Comment by nimbinatus** [comment]
@cnunciato I figured out how to get Hugo to spit out line numbers for code blocks, but right now it breaks the copy button:
- If I choose `inline`, the button copies the code *and* the line numbers.
- If I choose `table`, the line numbers get separated, but the button doesn't copy the code.

I'm guessing the copy button looks for a specific attribute. Think it's worth worrying about here in this PR? I was thinking the code was getting complex enough that it might have been worth it, but it's been fussy.

ref: https://gohugo.io/content-management/syntax-highlighting/#highlighting-in-code-fences

**Comment by nimbinatus** [comment]
:unsure-fry: I may need to re-rebase this...

**Comment by nimbinatus** [comment]
Blocked by pulumi/theme#145 as this pathway requires line numbers for reference points.

File: themes/default/content/learn/embedding-pulumi/building-api/index.md
Comment (aaronkao) [inline themes/default/content/learn/embedding-pulumi/building-api/index.md:2]: I think this title is a bit confusing because there are so many APIs associated with Pulumi that calling this a Pulumi API might not be getting your core point across

File: themes/default/content/learn/embedding-pulumi/getting-started/index.md
Comment (aaronkao) [inline themes/default/content/learn/embedding-pulumi/getting-started/index.md:24]: I would say we are now part of a platform engineering team and there are many developers from different business units building separate applications and this platform team's job is to build an interface/website/portal for people to request infrastructure

File: themes/default/content/learn/embedding-pulumi/getting-started/index.md
Comment (nimbinatus) [reply themes/default/content/learn/embedding-pulumi/getting-started/index.md:26]: The intent with our example for the pathway is to embed the Automation API program into a pipeline. You can use CI/CD to execute Pulumi programs, but under the hood my understanding is those integrations are Automation API that we've used to make it easier to integrate with those systems.

My intent here isn't to add every single use case, but rather to spark the imagination as the audience needs to think outside the box of the CLI. There's other pages and documents that lay out use cases. I don't want to go into the weeds.

File: themes/default/content/learn/embedding-pulumi/building-api/index.md
Comment (nimbinatus) [reply themes/default/content/learn/embedding-pulumi/building-api/index.md:2]: fixed

**Comment by nimbinatus** [comment]
Lint error being fixed in https://github.com/pulumi/pulumi-hugo/pull/1577

File: themes/default/content/learn/embedding-pulumi/getting-started/index.md
Comment (cnunciato) [inline themes/default/content/learn/embedding-pulumi/getting-started/index.md:38]: Should be `aws-python`?

```suggestion
Add `api` and `infra` directories in your project, and run `pulumi new` with the `aws-python` template in each one. If you need a refresher on how to do so, head to [Pulumi Fundamentals]({{< relref "/learn/pulumi-fundamentals" >}}). Add a directory called `time` inside your `api` directory to hold the actual application. Your project directory now should match this file tree:
```

File: themes/default/content/learn/embedding-pulumi/getting-started/index.md
Comment (cnunciato) [inline themes/default/content/learn/embedding-pulumi/getting-started/index.md:34]: What would you think of tweaking this phrasing a bit to play up the fact that we're running Pulumi programmatically ? (I ask because right now, it sounds a bit too much like we'll just be "shelling out" to Pulumi.)

```suggestion
We're going to embed Pulumi in a Python program so we can use it in pipelines, APIs, and web interfaces. For the Pulumi program we're using to test our automation, we'll get to some sample code in a moment. If you'd like to use your own Pulumi program, however, go for it!
```

File: themes/default/content/learn/embedding-pulumi/considerations/index.md
Comment (cnunciato) [inline themes/default/content/learn/embedding-pulumi/considerations/index.md:2]: Minor thing, but this reads like "extra considerations" are a kind of Pulumi-ism. I don't have a great suggestion for how to address it, but since it'll end up as the H1 of the page, and these pages are so well indexed, it might be a good place to use more relevant keywords, like "logging" or "exceptions", maybe.

File: themes/default/content/learn/embedding-pulumi/using-the-api/index.md
Comment (cnunciato) [inline themes/default/content/learn/embedding-pulumi/using-the-api/index.md:50]:
```suggestion
Now, we'll try `curl`ing the `location` endpoint. Open a new terminal window and run `curl localhost:8000/location`:
```

File: themes/default/content/learn/embedding-pulumi/using-the-api/index.md
Comment (cnunciato) [inline themes/default/content/learn/embedding-pulumi/using-the-api/index.md:161]:
```suggestion
What if you could migrate a database with Pulumi? It's possible! The examples in [Go](https://github.com/pulumi/automation-api-examples/blob/main/go/database_migration), [Python](https://github.com/pulumi/automation-api-examples/tree/main/python/database_migration), [TypeScript/Javascript](https://github.com/pulumi/automation-api-examples/blob/main/nodejs/databaseMigration-ts), and [C#/.NET](https://github.com/pulumi/automation-api-examples/blob/main/dotnet/DatabaseMigration) show a basic creation of a table with insertion and verification of data. If we combine this idea with the same type of idea of a web portal, we could make a plain migration tool that takes data from one table and generates a new one from that data, effectively migrating the table from one database to another.
```

File: themes/default/content/learn/embedding-pulumi/using-the-api/index.md
Comment (cnunciato) [inline themes/default/content/learn/embedding-pulumi/using-the-api/index.md:165]:
```suggestion
```

File: themes/default/layouts/shortcodes/code-filename.html
Comment (cnunciato) [inline themes/default/layouts/shortcodes/code-filename.html:3]: We have Tailwind helpers for these things, and should generally stick to them wherever we can, for overall design consistency. (Also threw in a structural tweak just to encapsulate things a bit more.)

```suggestion
<div class="border-solid border-t border-gray-300 pt-2">
    <h6 class="nohighlight">
        <span class="font-mono">{{ $file }}</span>
    </h6>
    <div>
        {{- .Inner | markdownify -}}
    </div>
</div>
```

**Review by cnunciato — APPROVED** [review]
Some suggestions you'll probably want to take a look at -- and I think you might want to remove that placeholder image -- looks good! Thanks for doing this. 👏

File: themes/default/content/learn/embedding-pulumi/getting-started/index.md
Comment (cnunciato) [reply themes/default/content/learn/embedding-pulumi/getting-started/index.md:74]: ☝️ I see a handful of places where it looks like we have multiple successive empty lines, so you might want to go through and tidy those up as well.

Comment (nimbinatus) [reply themes/default/content/learn/embedding-pulumi/getting-started/index.md:74]: Ah! Those are actually PEP8 style :) https://peps.python.org/pep-0008/#blank-lines

File: themes/default/content/learn/embedding-pulumi/getting-started/index.md
Comment (nimbinatus) [reply themes/default/content/learn/embedding-pulumi/getting-started/index.md:88]: Rejecting per PEP8

**Comment by nimbinatus** [comment]
Note: Dismissing Kat's review considering PTO and how far it's come.

File: themes/default/content/learn/embedding-pulumi/considerations/index.md
Comment (nimbinatus) [reply themes/default/content/learn/embedding-pulumi/considerations/index.md:2]: Hmm... good point.

**Comment by nimbinatus** [comment]
@cnunciato re:the placeholder images, we've been leaving them for each topic page because the system doesn't render without them (e.g., https://github.com/pulumi/pulumi-hugo/blob/master/themes/default/content/learn/building-with-pulumi/secrets/meta.png). Think we should find a way to remove them as a spike later?

File: themes/default/content/learn/embedding-pulumi/considerations/index.md
Comment (nimbinatus) [reply themes/default/content/learn/embedding-pulumi/considerations/index.md]: Fixed

**Pull request merged**
